### PR TITLE
Start implementing --head and --tail flags for sort.

### PR DESCRIPTION
**Pull request by shlomif**
See:

https://github.com/briandfoy/PerlPowerTools/pull/65

similar to "sort | head -10" or "sort | tail -10" but should
be more efficient (using a priority queue / heap).

Note that this is highly experimental, but comments are welcome.

----

I hereby disclaim any implicit or explicit ownership of my changes in this
changeset, and put them under a multiple licence consisting of your choice of
one of more of:

- The CC0 / Public Domain - https://creativecommons.org/choose/zero/ .

- The MIT / Expat license - https://en.wikipedia.org/wiki/MIT_License

- The default licence of your project

- The https://en.wikipedia.org/wiki/GNU_Lesser_General_Public_License - version
2.1 or higher

- The https://en.wikipedia.org/wiki/GNU_General_Public_License - version 2 or
higher

- Any licence in the 2018-Aug-27 popular licenses list of
https://opensource.org/licenses

- The https://en.wikipedia.org/wiki/Apache_License version 2.0 or later

- The https://en.wikipedia.org/wiki/Artistic_License version 2.0 or later

- The https://en.wikipedia.org/wiki/ISC_license

- The https://opensource.org/licenses/BSD-2-Clause

Crediting me will be nice, but not mandatory, and you can change the licence
of the project without needing my permission.

### CODE REVIEW

**Comment by shlomif** [comment]
Latest benchmarks on the 0fc branch:

```
+ set -e -x
+ inputfn=0fc-log.txt
+ sh=shuffled.txt
+ test -f shuffled.txt
+ cargo_dir=/home/shlomif/.cargo/bin
+ cargo_sort=/home/shlomif/.cargo/bin/moved-bin/sort
+ cargo_uu=/home/shlomif/.cargo/bin/uutils
+ test -e /home/shlomif/.cargo/bin/moved-bin/sort
+ _t /home/shlomif/.cargo/bin/moved-bin/sort -n --head 100
+ eval time /home/shlomif/.cargo/bin/moved-bin/sort -n --head 100
++ /home/shlomif/.cargo/bin/moved-bin/sort -n --head 100

real	0m23.422s
user	0m23.366s
sys	0m0.035s
+ _t sort -n '|' head -100
+ eval time sort -n '|' head -100
++ sort -n
++ head -100

real	0m9.407s
user	1m11.094s
sys	0m0.399s
+ _t /home/shlomif/.cargo/bin/moved-bin/sort -n --tail 100
+ eval time /home/shlomif/.cargo/bin/moved-bin/sort -n --tail 100
++ /home/shlomif/.cargo/bin/moved-bin/sort -n --tail 100

real	0m28.650s
user	0m28.598s
sys	0m0.026s
+ _t sort -n '|' tail -100
+ eval time sort -n '|' tail -100
++ sort -n
++ tail -100

real	0m15.943s
user	1m25.300s
sys	0m0.938s

```

**Comment by shlomif** [comment]
Hi! Does anyone have any comments about the features proposed in this PR? Would you be happy to merge a revised version (which remedies the file conflicts)? Ping!

**Comment by sylvestre** [comment]
I will have a look. Could you please rebase your patch ?

**Comment by shlomif** [comment]
> I will have a look. Could you please rebase your patch ?

Thanks!

I have rebased the branch. Should I also squash the commits? Sorry for the delayed reply.

**Comment by shlomif** [comment]
@sylvestre [and all]: ping!

**Comment by sylvestre** [comment]
Sorry, I didn't see the update!

Please reformat the patches (rustfmt) and fix the clippy warnings :)

**Comment by shlomif** [comment]
@sylvestre : hi! Hopefully it is OK now. I am using rustfmt and cargo-clippy from the fedora 32 x86-64 repositories.

**Comment by sylvestre** [comment]
Thanks, sorry again for the latency :)

**Comment by sylvestre** [comment]
Seems that your patch needs rebasing. I see all the recent history of coreutils in this PR :)

**Comment by shlomif** [comment]
@sylvestre : hi! I rebased with squashing commits and did a `git push --force`

File: src/uu/sort/src/sort.rs [hunk 1]
```diff
@@ -63,10 +65,44 @@ impl Default for Settings {
...
+impl<'a> PartialOrd for HeadHeapLine<'a> {
+    fn partial_cmp(&self, other: &HeadHeapLine) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl<'a> PartialEq for HeadHeapLine<'a> {
+    fn eq(&self, other: &HeadHeapLine) -> bool {
+        Ordering::Equal == compare_by(&self.current_line, &other.current_line, &self.settings)
+    }
```
Comment (sylvestre) [inline src/uu/sort/src/sort.rs:101]: Codecov isn't happy about coverage, could you please extend the tests?

File: src/uu/sort/src/sort.rs [hunk 2]
```diff
@@ -196,6 +246,13 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
         "Sort by SemVer version number, eg 1.12.2 > 1.1.2",
     );
     opts.optflag("c", "check", "check for sorted input; do not sort");
+    opts.optopt("", "head", "output a maximal number of elements", "INTEGER");
```
Comment (sylvestre) [inline src/uu/sort/src/sort.rs:249]: oh, if you are motivated, in a different PR, would be great to migrate this to clap :)

File: src/uu/sort/src/sort.rs [hunk 3]
```diff
@@ -239,6 +296,28 @@ With no FILE, or when FILE is -, read standard input.",
...
+            }
+        },
+        Some(s) => {
+            settings.head = s.parse().unwrap();
+            match matches.opt_str("tail") {
+                None => {}
+                Some(_) => {
+                    crash!(
+                        1,
+                        "Invalid options\n{}",
```
Comment (sylvestre) [inline src/uu/sort/src/sort.rs:314]:
```suggestion
                        "Conflicting options\n{}",
```

File: src/uu/sort/src/sort.rs [hunk 4]
```diff
@@ -276,7 +358,58 @@ With no FILE, or when FILE is -, read standard input.",
     exec(files, &settings)
 }
 
+fn exec_head(files: Vec<String>, settings: &Settings) -> i32 {
```
Comment (sylvestre) [inline src/uu/sort/src/sort.rs:361]: I know the main function is called "exec" but head is probably enough here as function name

File: src/uu/sort/src/sort.rs [hunk 5]
```diff
@@ -276,7 +358,58 @@ With no FILE, or when FILE is -, read standard input.",
...
+                        lines.heap.pop();
+                    }
+                } else {
+                    break;
+                }
+            }
+        }
+    }
+    let mut revlines = Vec::new();
+    while !lines.heap.is_empty() {
```
Comment (sylvestre) [inline src/uu/sort/src/sort.rs:394]: please add some comments what we are doing

File: tests/by-util/test_sort.rs [hunk 6]
```diff
@@ -30,6 +30,16 @@ fn test_numeric_unsorted_ints() {
...
+#[test]
+fn test_numeric_unsorted_head_ints() {
+    test_head_helper("numeric_unsorted_head_ints", "-n", "--head", "5");
+}
+
+#[test]
+fn test_numeric_unsorted_tail_ints() {
+    test_head_helper("numeric_unsorted_tail_ints", "-n", "--tail", "5");
+}
+
```
Comment (sylvestre) [inline tests/by-util/test_sort.rs:42]: please add a test with the conflicting options

File: tests/by-util/test_sort.rs [hunk 7]
```diff
@@ -30,6 +30,16 @@ fn test_numeric_unsorted_ints() {
     test_helper("numeric_unsorted_ints", "-n");
 }
```
Comment (sylvestre) [inline tests/by-util/test_sort.rs:32]: Please when head is larger than the number of line

**Review by sylvestre — COMMENTED** [review]
Looks great, a few comments :)

**Comment by shlomif** [comment]
@sylvestre : thanks for your commentary on my changes! I'll try to amend the PR soon.

**Comment by shlomif** [comment]
Thanks, @sylvestre ! I incorporated most of your suggestions. Please review at your leisure.

**Comment by shlomif** [comment]
Hi @sylvestre !

Sorry for the late reply.

> Actually, I thought you were implementing something already support by the GNU coreutils but this isn't the case:
> 
> ```
> $ sort --head 5 build.rs
> sort: unrecognized option '--head'
> ```
> 
> and
> 
> ```
> $ sort --tail 5 build.rs
> sort: unrecognized option '--tail'
> ```
> 
> I am really sorry, I should have noticed that earlier.
> This isn't the goal of this project. Currently, this project is to mirror was is done with the GNU coreutils, not add new options :(
> 
> Again, sorry for not noticing this earlier :(

To be frank, I am a little disappointed, but not too much so. My primary motivation with this changeset was to be a prototype or stepping stone for implementing it in GNU sort (after the initial patch for PerlPowerTools sort) and also to learn rust-lang better. I understand that rejecting changes that are considered out-of-scope of the project is natural (see ["Never Try to Please Everyone"](https://shlomifishswiki.branchable.com/Never_Try_to_Please_Everyone/) ) and accept that. Thanks for the review and comments, and sorry if I too wasted your time.

**Comment by sylvestre** [comment]
Sure, sorry again :(

There are plenty of options in rust/coreutils which aren't implemented yet.

**Pull request closed without merging**